### PR TITLE
[APK Adapter CP] Add application controller

### PR DESCRIPTION
**Pull request by Amila-Rukshan**
## Purpose

This is to add control plane Application CRD and it's controller code.

Existing directory structure of kubebuilder setup got changed to add multiple api groups (`cp` and `dp`) using https://kubebuilder.io/migration/multi-group.html.

I have added temporary `config.toml` and `log_config.toml` file to run the adapter. Please check the `README.md` file included.

### CODE REVIEW

File: adapter/cmd/adapter/conf/config.toml [hunk 1]
```diff
@@ -0,0 +1,92 @@
+[adapter]
```
Comment (AmaliMatharaarachchi) [inline adapter/cmd/adapter/conf/config.toml:1]: Shall we keep only what we require for now here and remove others? just remove from the config.toml would be sufficient for the moment, we need to remove unnecessary configs from the code(default_config.go, etc) gradually.

File: adapter/cmd/adapter/conf/resources/adapter/security/keystore/mg.key [hunk 2]
```diff
@@ -0,0 +1,28 @@
+-----BEGIN PRIVATE KEY-----
```
Comment (AmaliMatharaarachchi) [inline adapter/cmd/adapter/conf/resources/adapter/security/keystore/mg.key:1]: please generate new certs for the APK without copying from the old cc

File: adapter/internal/adapter/adapter.go [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+		grpcOptions = append(grpcOptions, grpc.Creds(
+			credentials.NewTLS(&tls.Config{
+				Certificates: []tls.Certificate{cert},
+				ClientAuth:   tls.RequireAndVerifyClientCert,
+				ClientCAs:    caCertPool,
+			}),
+		))
+	} else {
+		logger.LoggerMgw.Warn("failed to initiate the ssl context: ", err)
+		panic(err)
```
Comment (AmaliMatharaarachchi) [inline adapter/internal/adapter/adapter.go:97]: Feel it's better to add a blocking error log instead

File: adapter/internal/adapter/adapter.go [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+			})
+		}
+	}()
+}
+
+// Run starts the XDS server and Rest API server.
+func Run(conf *config.Config) {
+	sig := make(chan os.Signal, 2)
+	signal.Notify(sig, os.Interrupt)
+	// TODO: (VirajSalaka) Support the REST API Configuration via flags only if it is a valid requirement
```
Comment (AmaliMatharaarachchi) [inline adapter/internal/adapter/adapter.go:141]: please fix/remove old todos

File: adapter/internal/discovery/xds/server.go [hunk 5]
```diff
@@ -0,0 +1,213 @@
+/*
+ *  Copyright (c) 2020, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (AmaliMatharaarachchi) [inline adapter/internal/discovery/xds/server.go:2]:
```suggestion
 *  Copyright (c) 2022, WSO2 LLC. (http://www.wso2.org) All Rights Reserved.
```

File: adapter/internal/operator/apis/cp/v1alpha1/application_types.go [hunk 6]
```diff
@@ -0,0 +1,83 @@
+/*
+Copyright 2022.
```
Comment (AmaliMatharaarachchi) [inline adapter/internal/operator/apis/cp/v1alpha1/application_types.go:2]: please check and fix license headers

File: adapter/internal/operator/controllers/cp/application_controller.go [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+		applicationCache: applicationCache{},
+	}
+	c, err := controller.New("Application", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		loggers.LoggerAPKOperator.Errorf("Error creating Application controller: %v", err)
+		return err
+	}
+
+	if err := c.Watch(&source.Kind{Type: &cpv1alpha1.Application{}}, &handler.EnqueueRequestForObject{}); err != nil {
+		loggers.LoggerAPKOperator.Errorf("Error watching Application resources: %v", err)
```
Comment (AmaliMatharaarachchi) [inline adapter/internal/operator/controllers/cp/application_controller.go:63]: please add proper error logs with codes

File: adapter/internal/adapter/adapter.go (hunk 3 above)
Comment (Amila-Rukshan) [reply adapter/internal/adapter/adapter.go]: changed to Error

File: adapter/cmd/adapter/conf/resources/adapter/security/keystore/mg.key (hunk 2 above)
Comment (Amila-Rukshan) [reply adapter/cmd/adapter/conf/resources/adapter/security/keystore/mg.key:1]: Created this issue: https://github.com/wso2/apk/issues/146

File: adapter/internal/operator/controllers/cp/application_controller.go [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+			Severity:  logging.BLOCKER,
+			ErrorCode: 5001,
+		})
+		return err
+	}
+
+	if err := c.Watch(&source.Kind{Type: &cpv1alpha1.Application{}}, &handler.EnqueueRequestForObject{}); err != nil {
+		loggers.LoggerAPKOperator.ErrorC(logging.ErrorDetails{
+			Message:   fmt.Sprint("Error watching Application resources. ", err.Error()),
+			Severity:  logging.BLOCKER,
```
Comment (AmaliMatharaarachchi) [inline adapter/internal/operator/controllers/cp/application_controller.go:71]: also update the sheet https://docs.google.com/spreadsheets/d/1VH2rMZ-PdPbFw6VjUWV0o1Y0gxKANuVhgUpdYD1Iiog/edit#gid=189800195 with new codes

File: adapter/internal/operator/controllers/cp/application_controller.go [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+// Modify the Reconcile function to compare the state specified by
+// the Application object against the actual cluster state, and then
+// perform operations to make the cluster state reflect the state specified by
+// the user.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.13.0/pkg/reconcile
+func (r *ApplicationReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
```
Comment (AmaliMatharaarachchi) [inline adapter/internal/operator/controllers/cp/application_controller.go:94]: please use meaningful names. fix it in all changes
```suggestion
func (applicationReconciler *ApplicationReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
```

**Review by AmaliMatharaarachchi — APPROVED** [review]

**Pull request merged**